pkg/actor/resize_pvc.go

```diff
@@ -0,0 +1,222 @@
+		return err
+	}
+
+	return nil
+}
+
+// findAndResizePVC finds all active PVCs and resizes them to the new size contained in the cluster
+// definition.
+func (rp *resizePVC) findAndResizePVC(ctx context.Context, sts *appsv1.StatefulSet, cluster *resource.Cluster,
+	clientset *kubernetes.Clientset) error {
+	// K8s doesn't provide a way to tell if a PVC or PV is currently in use by
+	// a pod. However, it is safe to assume that any PVCs with an ordinal great
+	// than or equal to the sts' Replicas is not in use. As only pods with with
+	// an ordinal < Replicas will exist. Any PVCs with an ordinal less than
+	// Replicas is in use. To detect this, we build a map of PVCs that we
+	// consider to be in use and skip and PVCs that it contains
+	// the name of.
+	log := rp.log.WithValues("CrdbCluster", cluster.ObjectKey())
+	log.Info("starting finding and resizing all PVCs")
+	prefixes := make([]string, len(sts.Spec.VolumeClaimTemplates))
+	pvcsToKeep := make(map[string]bool, int(*sts.Spec.Replicas)*len(sts.Spec.VolumeClaimTemplates))
+	for j, pvct := range sts.Spec.VolumeClaimTemplates {
+		prefixes[j] = fmt.Sprintf("%s-%s-", pvct.Name, sts.Name)
+
+		for i := int32(0); i < *sts.Spec.Replicas; i++ {
+			name := fmt.Sprintf("%s-%s-%d", pvct.Name, sts.Name, i)
+			pvcsToKeep[name] = true
+		}
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(sts.Spec.Selector)
+	if err != nil {
+		return errors.Wrap(err, "converting statefulset selector to metav1 selector")
+	}
+
+	pvcs, err := clientset.CoreV1().PersistentVolumeClaims(cluster.Namespace()).List(ctx, metav1.ListOptions{
+		LabelSelector: selector.String(),
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "finding PVCs to for resizing")
+	}
+
+	for _, pvc := range pvcs.Items {
+		// Resize PVCs that are still in use
+		if pvcsToKeep[pvc.Name] {
+			pvc.Spec.Resources.Requests.Storage().Set(cluster.Spec().DataStore.VolumeClaim.PersistentVolumeClaimSpec.Resources.Requests.Storage().MilliValue())
+			if _, err := clientset.CoreV1().PersistentVolumeClaims(cluster.Namespace()).Update(ctx, &pvc, metav1.UpdateOptions{}); err != nil {
+				return errors.Wrap(err, "error resizing PVCs")
+			}
+		}
+	}
+
+	log.Info("found and resized all PVCs")
+
+	return nil
+}
+
+// rollSts performs a rolling update on the cluster.
+func (rp *resizePVC) rollSts(ctx context.Context, cluster *resource.Cluster, clientset *kubernetes.Clientset) error {
+
+	updateRoach := &update.UpdateRoach{
+		StsName:      cluster.StatefulSetName(),
+		StsNamespace: cluster.Namespace(),
+	}
+
+	podUpdateTimeout := 10 * time.Minute
+	podMaxPollingInterval := 30 * time.Minute
+	sleeper := update.NewSleeper(1 * time.Minute)
+
+	k8sCluster := &update.UpdateCluster{
+		Clientset:             clientset,
+		PodUpdateTimeout:      podUpdateTimeout,
+		PodMaxPollingInterval: podMaxPollingInterval,
+		Sleeper:               sleeper,
+	}
+
+	return update.RollingRestart(ctx, updateRoach, k8sCluster, rp.log)
```

Comment:
hey @chrislovecnm this feature is great!
but from v1.15, kubernetes feature `ExpandInUsePersistentVolumes` is in beta which doesn't require pods to be restarted. This avoids unnecessary disruption when expanding the disks. And is supported by GKE already for example. Can we make the restart optional and go with the option of only updating the sts definition? which we can achieve using a delete `--cascade=false` and recreating.

This would provide the necessary resilience that we expect from our cockroachDB clusters

Reply:
Thanks for the suggestion @zmalik! We can certainly look into adding support for `ExpandInUsePersistentVolumes`.

> This would provide the necessary resilience that we expect from our cockroachDB clusters

A rolling update should be safe as long as you have > 3 nodes in your CockroachDB cluster. Have you experienced otherwise? If so, I'd love to hear any additional details you might have to share!

Reply:
I do not think we need to make it optional, I would recommend using the beta feature, and only supporting 1.15 and up. @chrisseto thoughts?

Reply:
+1, it seems easier to implement and easier to use.

Reply:
awesome!

> A rolling update should be safe as long as you have > 3 nodes in your CockroachDB cluster

you are right but it still needs the connections to be rebalanced and I know we should be ready for any unexpected downtime of a node but if we can avoid it, its even better. And only use a restart and bounce the connections when there is actually no other way around.

Reply:
@zmalik so I tested this by hand and noticed a couple of things.  We cannot resize the statefulset directly. I am still getting the status that the pods have to be restarted, but that goes away.  Kinda weird.  The volume does resize.

Now, this is only true for some storage classes, so we do need the code for storage classes, that won't resize, and an API value to enable resizing without Pod restarts as well.

@chrisseto do you want another API value, or should the user restart the Pods?

1. Have API value that allows pods to skip restart on resize
2. Add code to not have statefulset volume template update
3. Modify checks in code that only tests the PVC size vs the cluster CR, because we cannot trust the statefulset
4. Figure out if need to modify the volumeclaimtemplate.

```bash
kubectl edit pvc <name> # for each PVC in the StatefulSet, to increase its capacity.
kubectl delete sts --cascade=false # <name> to delete the StatefulSet and leave its pods.
kubectl apply -f <name> # to recreate the StatefulSet.
kubectl rollout restart sts <name> # to restart the pods, one at a time. During restart, the pod's PVC will be resized.
```

Reply:
> so we do need the code for storage classes, that won't resize

@chrislovecnm right, that why I said to make it optional so we can introduce a flag like `restartPodsOnResize` which can be true by default, but users which have a storage-class which supports resizing without restart can set it to `true` 

to modify `volumeclaimtemplate` unfortunately there is no way right now except then recreating the sts https://github.com/kubernetes/kubernetes/issues/68737

we are deleting the sts using `cascade=false` and recreating it, which doesn't restart the pods either.

Reply:
@zmalik PTAL.